codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java

```diff
@@ -0,0 +1,403 @@
+                            });
+
+                            writer.openBlock("default:", "", () -> {
+                                writer.write("// Do nothing and ignore");
+                            });
+                        });
+                    });
+                    writer.write("*v = sv");
+                    writer.write("return nil");
+        });
+    }
+
+    private String getUnwrappedMapDelegateFunctionName(GenerationContext context, Shape shape) {
+        return ProtocolGenerator.getDocumentDeserializerFunctionName(shape, context.getProtocolName()) + "Unwrapped";
+    }
+
+    @Override
+    protected void deserializeStructure(GenerationContext context, StructureShape shape) {
+        GoWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(shape);
+
+        // Initialize the value now that the start stub has verified that there's something there.
+        writer.write("var sv $P", symbol);
+        writer.openBlock("if *v == nil {", "", () -> {
+            writer.write("sv = &$T{}", symbol);
+            writer.openBlock("} else {", "}", () -> {
+                writer.write("sv = *v");
+            });
+        });
+
+        // TODO: Refactor this to be more inline and DRY when handling xml attributes
+        Predicate<Shape> hasXmlAttributeTraitMember = shape1 -> {
+            Collection<MemberShape> members = shape1.members();
+            for (MemberShape member: members) {
+                if (member.hasTrait(XmlAttributeTrait.ID)) {
+                    return true;
+                }
+            }
+            return false;
+        };
+
+        if (hasXmlAttributeTraitMember.test(shape)) {
+            writer.openBlock("for _, attr := range decoder.StartEl.Attr {", "}", () -> {
+                writer.write("val := []byte(attr.Value)");
+                writer.openBlock("switch attr.Name.Local {", "}", () -> {
+                    Set<MemberShape> members = new TreeSet<>(shape.members());
+                    for (MemberShape member : members) {
+                        // check if member is not a document binding or does not have a xmlAttribute trait
+                        if (!memberFilter.test(member) || !member.hasTrait(XmlAttributeTrait.ID)) {
+                            continue;
+                        }
+
+                        String memberName = symbolProvider.toMemberName(member);
+                        String serializedMemberName = getSerializedMemberName(member);
+                        writer.openBlock("case $S:", "", serializedMemberName, () -> {
+                            String dest = "sv." + memberName;
+                            context.getModel().expectShape(member.getTarget()).accept(
+                                    getMemberDeserVisitor(member, dest, ((shape1) -> shape1.hasTrait(XmlAttributeTrait.ID))));
```

Comment:
This is testing the target shape, which cannot have the xml attribute trait since it only applies to members.